File: domain/src/main/java/org/openmedstack/domain/Aggregate.java
package org.openmedstack.domain;

import openmedstack.events.DomainEvent;

import java.util.List;

public interface Aggregate {
    String getId();
    int getVersion();
    void applyEvent(Object event) throws HandlerForDomainEventNotFoundException;
    List<DomainEvent> getUncommittedEvents();
    void clearUncommittedEvents();
    Memento getSnapshot();
}

//public interface IDetectConflicts
//     Boolean ConflictDelegate<TUncommitted extends BaseEvent,TCommitted extends BaseEvent >(
//        TUncommitted uncommitted,
//        TCommittedcommitted)
//        where TUncommitted : class where TCommitted : class;

File: modules/domain/src/main/kotlin/org/openmedstack/domain/Aggregate.kt
package org.openmedstack.domain

import org.openmedstack.events.DomainEvent

interface Aggregate {
    val id: String
    val version: Int
    @Throws(HandlerForDomainEventNotFoundException::class)
    fun applyEvent(event: DomainEvent)
    val uncommittedEvents: MutableList<DomainEvent>
    fun clearUncommittedEvents()
    fun getSnapshot(): Memento
}
